tesseract-server/src/auth.rs

```diff
@@ -55,21 +31,76 @@ struct Claims {
     sub: String,
     status: String,
     exp: usize,
+    auth_level: Option<i32>,
 }
 
+pub fn extract_token(req: &HttpRequest<AppState>) -> String {
+    let qry = req.query();
+
+    let token = {
+        let qp_token = qry.get(X_TESSERACT_JWT_TOKEN);
+        match qp_token {
+            None => {
+                // If we don't match in query params, try headers
+                // The next lines below are little ugly. Basically,
+                // we need to catch for two potential errors:
+                // 1. the key might not be present (phase1)
+                // 2. the key might not parse to a string properly (phase2)
+                let phase1 = req.headers().get(X_TESSERACT_JWT_TOKEN);
+                match phase1 {
+                    Some(val) => {
+                        let phase2 = val.to_str();
+                        match phase2 {
+                            Ok(v) => v,
+                            _ => ""
+                        }
+                    },
+                    _ => "",
+                }
+            },
+            Some(token) => token,
+        }
+    };
+    token.to_string()
+}
 
-pub fn validate_web_token(jwt_secret: &Option<String>, raw_token: &str) -> bool {
+// None = auth not set on server, -1 = bad auth level
+pub fn user_auth_level(jwt_secret: &Option<String>, raw_token: &str) -> Option<i32> {
     match jwt_secret {
         Some(key) => {
-            let validation = Validation {
-                ..Validation::default()
-            };
+            let validation = Validation::default();
             match decode::<Claims>(&raw_token, key.as_ref(), &validation) {
                 Ok(c) => {
                     let claims: Claims = c.claims;
-                    claims.status == "valid" // TODO allow this value to be configurable
+                    let part1 = claims.status == "valid";
+                    if part1 {
+                        Some(claims.auth_level.unwrap_or(0)) // default auth level 0
+                    } else {
+                        Some(0)
+                    }
                 },
-                Err(_) => false, // If any error occurs, do not validate
+                Err(_err) => Some(-1), // If any error occurs, do not validate
+            }
+        },
+        None => None
+    }
+}
+
+pub fn validate_web_token(jwt_secret: &Option<String>, raw_token: &str, min_auth_level: i32) -> bool {
+    match jwt_secret {
+        Some(key) => {
+            let validation = Validation::default();
+            match decode::<Claims>(&raw_token, key.as_ref(), &validation) {
+                Ok(c) => {
+                    let claims: Claims = c.claims;
+                    let part1 = claims.status == "valid"; // TODO allow this value to be configurable
+                    let part2 = match claims.auth_level {
+                        Some(lvl) => lvl >= min_auth_level,
+                        None => true
```

Comment:
Should we require that an `auth_level` be provided? I believe that right now if no level was included in the JWT, we'd allow cube access even if that cube had a min auth level > 0, no?

Reply:
At first I was thinking it would be easier to allow default access level if no auth level was specified to be 0 but I think it in terms of security probably better to err on the side of requiring more specificity than less, so I'll change this